src/transports/smart_protocol.c

```diff
@@ -395,16 +397,13 @@ int git_smart__negotiate_fetch(git_transport *transport, git_repository *repo, c
 				if ((error = store_common(t)) < 0)
 					goto on_error;
 			} else {
-				pkt_type = recv_pkt(NULL, buf);
-
-				if (pkt_type == GIT_PKT_ACK) {
+				error = recv_pkt(NULL, &pkt_type, buf);
+				if (error < 0) {
+					goto on_error;
```

Comment:
I'd propose to make it `if ((error = recv_pkt(NULL, &pkt_type, buf)) < 0) goto on_error;`. This would further disentangle handling of errors and the packet type.